### DISCUSSION TITLE
Sentence Spacing (Double Spacing)

### DISCUSSION THREAD

**Discussion by bielok** [typst/typst#4403] [Q&A] [answered]
Is it possible to have configurable intra-paragraph sentence spacing? For example, to achieve double spacing after a period. If so, how?

This is purely stylistic as I find it easier to skip forward while skimming.

Kind regards.

**Comment by Mc-Zen** [comment, marked as answer]
Hi @bielok 
This is indeed not so straightforward to get right. I just experimented a few minutes but let's look at some options. 

You could use text `show` rules like this:
```typ
#show ". ": "." + h(1em, weak: true)
```
This transforms every `". "` into a point followed by a configurable horizontal space. It needs to be weak, so that consecutive space characters are swallowed and inconsistent spacing is avoided. This approach looks nice at first sight but has a problem. If a line in a paragraph ends with a period and justification is turned on, then this `h() `space will be shown so that the period does not align with the right (assuming left-to-right text) edge. 

Replacing a period and one space by a period followed by two or three spaces solves this and looks much better:
```typ
#show ". ": ".   "
```
Note that we match the space after the period, i.e., `#show ". "` instead of `#show "."`. This prevents matching (at least some) punctuation used for example together with numbering of headings, equations etc...

This `show` rule might still apply to some false positives, however, it's hard to get past that currently (with no negating selectors available).

**Reply by bielok** [reply]
Understood, this will also affect -- common -- instances of `i.e.` or `e.g.` and perhaps `etc.`.
However, for `#show ". ": ".  "` it suffices to add one extra space after the false positive, for example, `lorem i.e.  ipsum`.

I know you already know this, just leaving it here in case anyone else is interested.

This alternative is good enough for me, thanks for taking the time to figure it out!

**Reply by Mc-Zen** [reply]
You're welcome. And good that you pointed that one out!

**Reply by laurmaedje** [reply]
The false positive fix with `lorem i.e.  ipsum` relies on buggy behaviour (https://github.com/typst/typst/issues/86), so it might break in the future.

**Reply by bielok** [reply]
Two alternative solutions I can think of:

#### Using math
```
Lorem ipsum $"i.e."$ dolor sit amet.
```

#### Using a let binding

```
#let i-e = "i.e."
Lorem ipsum #i-e dolor sit amet.
```

**Reply by laurmaedje** [reply]
Let binding would also stop working if the bug is fixed. A more idiomatic thing might be using a non breaking space after i.e. with ~ (which might be a good idea in general?)

**Reply by bielok** [reply]
How about multiple show rules?

```
#show ". ": ".  "
#show "i.e.": "i.e."
#show "e.g.": "e.g."
#show "etc.": "etc."

Lorem i.e. ipsum. // No double space
```

This should match a single text node every time, correct? Tested it and currently works, but relies on the correct declaration order of the rules, which I hope is fine.

---

Also, I am not sure how the non-breaking space solution works, could you show me a snippet? 

```
#show ". ": ".  "
Lorem etc.~ ipsum.
```
Still produces a false positive.

**Reply by Mc-Zen** [reply]
> This should match a single text node every time, correct? Tested it and currently works, but relies on the correct declaration order of the rules, which I hope is fine.

As Laurenz pointed out, this currently works because text show rules don't work acroess node boundaries. But this will be hopefully changed (because it's a bit limiting). 

> Still produces a false positive.

Actually it doesn't! You only get the double space because you have the non-breaking space followed by a normal space. You need to write
```
#show ". ": ".  "
Lorem etc.~ipsum.
```
Question remains, if you really want to forbid breaking after every "i.e."

**Reply by Mc-Zen** [reply]
You can also consider doing something like
```
#show ".. ": ".  "
Lorem etc. ipsum.. Dolor sit amet.. 
```

It does not look nice in the source code but this is much more fool-proof and has considerably fewer side effects.

**Reply by bielok** [reply]
Thanks to both of you!

**Reply by robamler** [reply]
I know I'm late to the party, but for other googlers who arrive here, I came up with the following variation of this proposal:

```typst
#show ". ": ". " + h(0.15em, weak: true)
#show ": ": ": " + h(0.15em, weak: true)
#show "; ": "; " + h(0.15em, weak: true)
```

**Rationale:**
- Two spaces looked too wide to me, so I instead went for a single space character plus 0.15em of extra horizontal spacing. This doesn't seem to cause any issues: I couldn't reproduce the issue with a space at the end of a line of a `justify`ed paragraph. Even if I increase the width of the `h` significantly, any lines that happen do end in a full stop still perfectly align at the right edge if `justify` is set to `true`, and the next line still perfectly aligns at the left edge too (I'm using typst 0.12.0).
- I went for a space character _and_ an `h` function rather than just a single wider `h` because `h` does not get stretched in `justify`ed paragraphs like space characters do. Thus, if you use only a wide `h` and no space character for the space between sentences, then that space can sometimes end up _narrower_ than the normal spaces between words in lines that happen to get stretched a lot.
- It feels sensible to me to have the same effect for ":" and ";" as well.

**Aside:** "e.g." and "i.e." don't pose an issue in American English since most style guides recommend using a comma after these anyway (British style guides don't, however).
